### DISCUSSION TITLE
Update a static page by event

### DISCUSSION THREAD

**Discussion by MohammadPCh** [vercel/next.js#16488] [Help] [answered]
# Feature request
Provide an API to update a static page not all of them.

## Problem
NEXTJS introduced Incremental Static Re-generation mechanism to update existing pages, but it's useless for cases like this:
consider you have a website about movies, with this the general URL is:
**https://example.com/movies/[id]**
your website has about 200K movies and they build in build time or run time. so you build about 400K files (.json & .html) on your disk. your movies page doesn't change continuously, maybe a typo occurred or site visitors add comments then your page should update.
Now how we can update only that page and avoid build all 200K pages again? Is it clear? let me give you an example:
NextJs build all 200K page and we want to change the description of a movie with id 60, so... What should we do?
We have 3 ways:
1. build all the pages again! (worst idea! build 200K pages just for one page)
2. use Incremental Static Re-generation (worse idea! because it builds all the page continuously but change frequency of movie page is constant maybe change every second may change once in a year).
3. remove 60.html and 60.json file from build folder and restart the nextjs server! (a bad idea! we need to restart nextjs because it serves files from memory);
all of these ways are inefficient and useless.

## Suggested solution

Nextjs needs to provide an API to get a route and update/remove it.
for example:

```
import { ssgUpdate, ssgPurge } from "next";
/*
.
.
.

*/
// for update
ssgUpdate('movies/[id]','movies/60').then(res=>{
    console.log('successful')
}).catch(err=>{
    console.log(err)
});

//for remove
ssgPurge('movies/[id]','movies/60').then(res=>{
    console.log('successful')
}).catch(err=>{
    console.log(err)
});

```
If ssgPurge called that page files removed from disk and memory;
If ssgUpdate called that page regenerates and if it's successful it updates memory and disk;

Thanks

**Comment by mdehghani** [comment]
Same issue here.
We want to update a specific static page in case of occurrence of an event (e.g. when a specific message is gotten on the message broker), but there isn't any solution to do that. The suggested solution solves our problem too 👍

**Comment by ysvahn** [comment]
In the case of ISR, Next.js does not rebuild every page every time a change occurs, it only rebuilds the page when a user navigates to it and only if the page changed.

When would your functions be called ? ISR seems to do exactly the same without having to manually call these functions.

**Comment by MohammadPCh** [comment]
Hi @Kroonax 
There is difference between what i want and you solution!
consider you have static page about movie. the sample URL is:
https://example.com/1/god-father
This page has about 1 million visitors in day. 
i want to update it once when an event occurs for example when page received a comment. not all the time.
in your solution the page builds many times! but i need only once regeneration.
and thought about situation that you have about 100 thousands movie pages.

**Comment by devklepacki** [comment]
My understanding of this feature is that it regenerates only after a page changed. So on each request page is checked for newer version (with comment in your example). User is served a stale version and newer is build in background. Next time any other user requests it, the newer version is served statically each time, until there is again newer version.

**Comment by MohammadPCh** [comment]
Hi @pklepacki ,
Why we need to wait for a request?
When a comment received your back-end knows that this page should change. it's exactly the time that we can build new page and we don't need to wait someone request this page and then build it.

**Comment by tam315** [comment]
@MohammadPCh 
Hi, I found following blog.
I hope that these features will be comming.

https://nextjs.org/blog/next-9-5#stable-incremental-static-regeneration

> Up next, we will be working on a supplemental RFC to address two additional incremental static generation capabilities:
> - Re-generating and invalidating multiple pages at once (like your blog index and a certain blog post)
> - **Re-generating by listening to events** (like CMS webhooks), ahead of user traffic

**Reply by michrome** [reply]
Cross linking this to https://github.com/vercel/next.js/discussions/16112 which is a request for the same: regeneration based on an event, to cut short the wait for a revalidation period.

**Comment by MohammadPCh** [comment]
@junkboy0315 
Thank Shota Tamura,
I handle it with SSR and my own cache :)
If you interested i can describe how i handle it.
Good luck.

**Reply by techstarterau** [reply]
Hi Mohammad - I would be VERY interested to hear how you resolved this. Been waiting every day for a solution, I'm ready to just hack together anything that works.

**Reply by lorenzosignoretti** [reply]
Hi @MohammadPCh, +1. Would you be willing to share your solution?

**Comment by rodbs** [comment]
Same here. I'd love to update pages based on listening to events (CMS webhooks)

**Comment by Nick-Mazuk** [comment]
It would be absolutely great to have this, it would solve a ton of headaches and tradeoffs on my sites.

**Reply by lorenzosignoretti** [reply]
How are you currently handling the issue? Did you find any useful workaround?

**Comment by FSaldanha** [comment]
Here's a provisional solution: #22384

**Comment by antonioberetini** [comment]
Same here this is a huge gap from feature perspective specially if you pair Next with a headless CMS where content folks often modify and create new content.

**Comment by kristijorgji** [comment]
Hi, looking for this feature too.
Any updates or plans to have this ?

**Reply by kristijorgji** [reply]
In the meanwhile would like to share my current non official solution to this , while waiting for vercel team to officially provide a proper hook system.

I found that just deleting the page statically generated data (or via isr) will force next to re-render it

 ```
rm .next/server/pages/${locale}/posts/${postId}.json
rm .next/server/pages/${locale}/posts/${postId}.html
``` 

on next request will force next to rebuild it with latest data 🗡️ 😄  and works like expected.

It is working fine for a pet project, where only one web server is serving all requests.
For distributed load with multiple servers, can still build something forcing all servers to delete this particular page data but would rather wait and look forward to next official solution to this.

IMO they can easily provide a hook for it, or cli command like `next invalidate url` that does something similar as I am doing now on redis subscribe event (post update event)

**Reply by FSaldanha** [reply]
Besides these static files, data is saved in the internal cache. You should purge it too to make sure you won't serve stale content (see #22384)

**Reply by kristijorgji** [reply]
@FSaldanha thnx for info, how can I do that ? Will check linked issue
Is there any way to do that without using custom server.js (also would prefer to have server.ts in our project)

**Comment by kokoskiwi** [comment]
we really need that feature
https://www.reddit.com/r/nextjs/comments/q0rla1/can_i_tell_nextjs_if_a_site_needs_to_get/

**Comment by alexandrosk** [comment]
This is really important, not sure why its not stressed our aalready! 
I can only clear the cache when rebuilding the product -.-

**Comment by antonioberetini** [comment]
Maybe automatically hitting the URL when content changes will do the trick 😕

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by antonioberetini** [comment]
@enes-get-local sounds like a good challenge for you 😜 to work on

**Comment by meotimdihia** [comment]
Next.js's developer promised he won't forget this problem:
https://github.com/vercel/next.js/discussions/11552#discussioncomment-646963
We have to wait for it.

>leerob 23 days ago
> No update - I promise I'm not ignoring y'all, it's just still being worked on! What we originally thought was a good option took a bit more iteration. We are still really excited about this and want to improve ISR further. The reason for the delay isn't due to lack of effort, but just ensure we get the APIs and design correct 👍 Expect a public RFC for this, too.

**Reply by antonioberetini** [reply]
A year before and it is still not here... Sure they are working on it

**Comment by leerob** [comment, marked as answer]
I'm incredibly excited to share that Next.js 12.1 was just released, which includes **on-demand Incremental Static Regeneration**! We know many of you have been looking forward to this for a while, including myself, and I'm so happy it is here. 

- [Learn about Next.js 12.1](https://nextjs.org/blog/next-12-1)
- [Watch a demo of on-demand ISR](https://youtu.be/BGexHR1tuOA)
- [View another demo and give us your feedback](https://on-demand-isr.vercel.app/)
- [Read the documentation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta)

**Reply by demonhunter3333** [reply]
How do we update pages on dynamic routes with on-demand ISR?

**Reply by lorenzosignoretti** [reply]
@demonhunter3333 Depending on what you need to do, a query param on your revalidation API route can do the trick (see [this gist](https://gist.github.com/lorenzosignoretti/a457be6e25fd48e81e380b2bf689a025#file-revalidate-js)). More [here](https://medium.com/@lorenzosignoretti/how-does-next-js-new-on-demand-incremental-static-regeneration-work-f63b9c6c30ea).

**Reply by demonhunter3333** [reply]
> @demonhunter3333 Depending on what you need to do, a query param on your revalidation API route can do the trick (see [this gist](https://gist.github.com/lorenzosignoretti/a457be6e25fd48e81e380b2bf689a025#file-revalidate-js)). More [here](https://medium.com/@lorenzosignoretti/how-does-next-js-new-on-demand-incremental-static-regeneration-work-f63b9c6c30ea).

Thank you!

**Reply by kristijorgji** [reply]
@demonhunter3333 
Your blog post is not correct I think the unstable revalidate does more then just calling fetching page.
It ignores revalidate prop which is very good and excepted behavior from me.
If you make direct call to the page, and have let say revalidate: 120 but only 30 seconds passed you don't refresh cache

**Reply by lorenzosignoretti** [reply]
> @demonhunter3333 
> 
> Your blog post is not correct I think the unstable revalidate does more then just calling fetching page.
> 
> It ignores revalidate prop which is very good and excepted behavior from me.
> 
> If you make direct call to the page, and have let say revalidate: 120 but only 30 seconds passed you don't refresh cache

You're right, I'll clarify that. Thanks for pointing this out
